### ISSUE TITLE
BigInt and Float support

### ISSUE THREAD

**Issue by jimscarver** [rchain/rchip-proposals#51]
## Introduction/Motivation/Abstract
Data limitations create bugs.  Without BigInt we are forever plagued by overflow and underflow errors. One may blow up the world. without these types we are unable to marshal objects of those types faithfully using inter blockchain protocol. 

## Motivating Examples
@dckc pointed out these issues and how agoric contracts convert easily to rholang and there is a lot there we can use ERTP, IBC, CapTP, etc. However when BigInts and Float are used we will introduce bugs as already happened with the POS contract and again recently. See video. https://youtu.be/IbW6uvgCSv4?t=5463 

## Design
BigInt requires virtually a global change in the rnode source of Int to BigInt.

Adding floating point operations and constants is faily mechanical and requires an update to the rholang spec.

**Comment by dckc** [comment]
I suggest making floats a separate issue; those are much, much less important.

**Comment by Bill-Kunj** [comment]
> I suggest making floats a separate issue; those are much, much less important.

I agree. Arbitrary precision floats are far less important. 
Making BigInt universal is vital, as recent problems with PoS demonstrate. 
The fact that custom tooling has been required to support the total RPC token supply should be ample evidence.

**Comment by Bill-Kunj** [comment]
As a general reply to the abstract of this rchip, `data limitations create bugs` is more properly expressed as `data limitations ARE bugs`.

**Comment by jimscarver** [comment]
Implemented the **vital half — universal BigInt** — in the Rust node (rchain-community/rchain-rust#49, merged).

Integer arithmetic can no longer overflow or underflow:

- `Int` operations are checked; on `i64` overflow **both operands promote to `BigInt`** and the result is exact (previously `wrapping_add/sub/mul`, `wrapping_neg`).
- Mixed `Int`/`BigInt` operands now work (was a type error), comparisons span both forms, and `2 == 2n`.
- `i64::MIN / -1` and `i64::MIN % -1` are exact (2^63 and 0) instead of errors.
- Division/modulo by zero is still an error in both forms.

Values keep the `i64` representation while they fit, so the rest of the node (proto, hashing, canonical ordering, costs, system processes) is unchanged — the `i64` fast path charges the existing per-op cost, the promoted path charges `big_int_*`.

**This is a hard fork:** previously-wrapped results change, so prior state and block hashes are invalid — recorded in `spec/AUDIT.md` §6 (Scala-deviation register). Agreed it must land before any network runs.

**Float is not included.** Both comments here call floats far less important and suggest splitting them out; a float also needs a canonical total order and equality for the tuple space (NaN, ±0) plus a new `Expr`/proto variant that changes the state hash again. Suggestion: open a dedicated issue for floats and I will implement it next.

Also found while doing this: `rholang/src/merging.rs::calculate_num_channel_diff` computes `end_val.wrapping_sub(prev)` on `i64` channel values — the same silent-wrap class, on the merge path.

**Comment by jimscarver** [comment]
Split the float half out as #58 (with the design questions: canonical total order/equality for the tuple space, proto/sort change, mixed Int/Float arithmetic, costs, determinism). The BigInt half is implemented in rchain-community/rchain-rust#49.
